[WIP] Improve the pretty printer formatting for higher dimensional arrays

Abhiroop opened the pull request:
Attempt to fix https://github.com/diku-dk/futhark/issues/637

The pretty printer after this commit:

```
replicate 10 (replicate 10 0)
[[0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]
 ]
```
And for cases when the total columns exceeds 80 characters

```
[[0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 , [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32,
    0i32]
 ]
```

I played around with some of the other combinators of `mainland-pretty` but in most other cases the indentation gets spoiled. This could serve as a stopgap solution to the existing formatting. Another option could be to `flatten` the entire representation. This looks better to me. It works fairly well for higher dimensional arrays:

```
[9]> replicate 5 (replicate 5 (replicate 5 0))
[[[0i32, 0i32, 0i32, 0i32, 0i32]
  , [0i32, 0i32, 0i32, 0i32, 0i32]
  , [0i32, 0i32, 0i32, 0i32, 0i32]
  , [0i32, 0i32, 0i32, 0i32, 0i32]
  , [0i32, 0i32, 0i32, 0i32, 0i32]
  ]
 , [[0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    ]
 , [[0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    ]
 , [[0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    ]
 , [[0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    , [0i32, 0i32, 0i32, 0i32, 0i32]
    ]
 ]
```

In other array libraries from other languages like NumPy, the pretty printers are highly configurable: https://docs.scipy.org/doc/numpy/reference/generated/numpy.set_printoptions.html which could serve as an inspiration.

athas commented:
I'm not sure I like this.  While it solves the problem with the weirdest linebreaks, the leading commas are not particularly nice, since the columns end up being unaligned.

Abhiroop commented:
Closing this for now. If I can come up with something better I will reopen the PR. Most prettyprinters will generally intercalate with a leading comma, it might be easier to write the instance by hand.

Abhiroop commented:
Currently it renders something like this:

```
[3]> replicate 10 (replicate 10 0)
[[0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32],
 [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32],
 [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32],
 [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32],
 [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32],
 [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32],
 [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32],
 [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32],
 [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32],
 [0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32, 0i32]]
```

However it doesn't obey the 80 character per line limit and overflows till the end of the screen. 

@athas looking at the code, do you see if I am missing something obvious, to enforce the 80 characters per line limit?

Abhiroop commented:
Fixed the 80 character per line issue and squashed all my commits.

For higher dimensional arrays:

```
[2]> replicate 5 (replicate 5 (replicate 5 0))
[[[0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32]],
 [[0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32]],
 [[0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32]],
 [[0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32]],
 [[0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32],
  [0i32, 0i32, 0i32, 0i32, 0i32]]]
```

athas commented:
If you fix the style issues pointed out by Travis, then this is fine with me.  Definitely an improvement over the current situation.  I don't think the line wrapping is particularly important (but the reason you don't get it is because you use literal newlines and spaces instead of the combinators exposed by mainland-pretty).

athas commented:
Oh, didn't see you already fixed the combinator issue.  Travis is still upset about some pedantic `hlint` things, though.

Abhiroop commented:
@athas I have fixed the `hlint` complaints. There seems to be some general build failure in Travis.

athas commented:
Yeah, Travis sometimes stalls. I've given it a kick.
